### PR TITLE
Only null and undefined are valid values for same shape optimization

### PR DESCRIPTION
**Pull request by mcollina**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
This broke fastify-autoload tests.

File: test/decorator.test.js [hunk 1]
```diff
@@ -939,3 +939,32 @@ test('plugin required decorators', async t => {
   app.register(plugin2)
   await app.ready()
 })
+
+test('decorateRequest/decorateReply empty string', t => {
+  t.plan(7)
+  const fastify = Fastify()
+
+  fastify.decorateRequest('test', '')
```
Comment (jsumners) [inline test/decorator.test.js:947]: Can you add a test where the initial decoration is set to `undefined`?

**Review by jsumners — CHANGES_REQUESTED** [review]

File: test/decorator.test.js (hunk 1 above)
Comment (mcollina) [reply test/decorator.test.js:947]: I have added the test.

File: test/decorator.test.js [hunk 2]
```diff
@@ -939,3 +939,90 @@ test('plugin required decorators', async t => {
...
+
+test('decorateRequest/decorateReply is undefined', t => {
+  t.plan(7)
+  const fastify = Fastify()
+
+  fastify.decorateRequest('test', undefined)
+  fastify.decorateReply('test2', undefined)
+  fastify.get('/yes', (req, reply) => {
+    t.equal(req.test, null)
+    t.equal(reply.test2, null)
```
Comment (mcollina) [inline test/decorator.test.js:980]: @jsumners as you see the values are casted to `null`.

**Review by jsumners — APPROVED** [review]
Looks good, but can't say for certain this resolves my find. If we put out a rc release I can check easily.

**Comment by mcollina** [comment]
Releases numbers are cheap, I'll push a minor out. Our fastify-autoload breaks without this anyway.

**Review by Eomm — APPROVED** [review]

**Comment by jsumners** [comment]
Will test as soon as I can.

**Pull request merged**